In [56]:
import math
import numpy as np

### Chapter 3

#### As per the book's assumption, we assume that we have a black box that gives a random number upon request (which is np.random.rand() in this case)

In [54]:
def simulate_integral_value(k, h):
    ans = 0
    for _ in range(k):
        y = np.random.rand() # U[0, 1]
        ans += h(y)
    ans = ans/k
    return ans

#### Q6

In [175]:
# k = int(1e6)
# def h(y):
#     nmr = 1/y - 1
#     dnm = ((1+(1/y-1)**2)*y)**2
#     return nmr/dnm
# ans = 0
# for i in range(k):
#     y = np.random.rand()
#     ans += h(y)
# ans = ans/k
# print(ans)

In [60]:
k = int(1e6)
def h(y):
    nmr = 1/y - 1
    dnm = ((1+(1/y-1)**2)*y)**2
    return nmr/dnm
ans = simulate_integral_value(k, h)
print(ans)

0.49995430528505924


##### Matches with the exact answer

#### Q7

In [65]:
k = int(1e6)
def h(y):
    nmr = math.exp(-(1/y - 1)**2)
    dnm = y**2
    return nmr/dnm
ans = 2*simulate_integral_value(k, h)
print(ans)

1.771241916712693


##### Matches with the exact answer

#### Q9

In [68]:
def simulate_integral_value_2d(k, h):
    ans = 0
    for _ in range(k):
        u = np.random.rand() # U[0, 1]
        v = np.random.rand() # U[0, 1]
        ans += h(u, v)
    ans = ans/k
    return ans

In [71]:
k = int(1e6)
def h(u, v):
    if(u>=v):
        return 0
    return math.exp(-(1/u+1/v-2))/((u*v)**2)
ans = simulate_integral_value_2d(k, h)
print(ans)

0.49952046543913003


##### Matches with the exact answer

### Q10

In [76]:
k = int(1e6)
t1, t2, t3 = 0, 0, 0
for _ in range(k):
    u = np.random.rand()
    t1 += u*math.exp(u)
    t2 += u
    t3 += math.exp(u)
t1, t2, t3 = t1/k, t2/k, t3/k
ans = t1-t2*t3
print(ans)

0.1407680176628976


##### Matches with the exact answer

#### Q12 (a, b, c)

In [77]:
kvec = [100, 1000, 10000]
for k in kvec:
    for _ in range(k):
        cnt = 0
        val = 0
        while(val<=1):
            u = np.random.rand()
            cnt += 1
            val += u
        ans += cnt
    ans = ans/k
    print(k, ans)

100 2.661407680176629
1000 2.6886614076801765
10000 2.723468866140768


#### Q14

In [191]:
x = np.zeros(15)
x[1], x[2] = 23, 66
for i in range(3, 15):
    x[i] = (3*x[i-1]+5*x[i-2])%100
u = x/100
random_number_sequence = u[1:]
print(u[1:])

[0.23 0.66 0.13 0.69 0.72 0.61 0.43 0.34 0.17 0.21 0.48 0.49 0.87 0.06]


## Chapter 4

### Q1

In [29]:
# Program to generate n values from the pmf with p1 = 1/3 and p2 = 2/3
p1 = 1/3
def simulate_pmf12(n):
    output = []
    v = np.random.rand(n)
    for val in v:
        if(val<=p1):
            output.append(1)
        else:
            output.append(2)
    output = np.array(output)
    return output

In [30]:
for n in [100, 1000, 10000]:
    output = simulate_pmf12(n)
    assert(len(output) == n)
    prop = len(output[output == 1])/len(output)
    print(n, prop)

100 0.33
1000 0.324
10000 0.326


### Q2

In [31]:
def simulate_pmf_n(pvec) -> float:
    '''
    Args:
        pvec: vector of probabilities defining the probability mass function (pmf)
    Returns:
        value of the Random Variable following the given pmf
    '''
    n = len(pvec)
    v = np.random.rand()
    # The cdf value (upto index i; i ranging from 1 to n)
    F = p1
    ans = -1
    for i in range(1, n+1):
        if(v<=F):
            ans = i
            break
        if(i==n):
            raise Exception("index of iteration should not reach n!")
        F += pvec[i]
    assert(ans != -1)
    return ans

### Q3

In [233]:
u = np.random.rand()
if(u<=0.35):
    ans = 3
elif(u<=0.65):
    ans = 1
elif(u<=0.85):
    ans = 2
else: # u will surely be <= 1
    ans = 4
print(ans)

3


### Q4

In [107]:
n = 100
v = []

num_trials = 50000
for _ in range(num_trials):
    # seq will represent the sequence of shuffled cards

    # Creating a base sequence
    seq = np.array([i for i in range(n+1)])
    
    # Shuffling the sequence
    for k in range(n, 1, -1):
        u = np.random.rand()
        index = int(k*u) + 1
        
        # swapping seq[index] and seq[k]
        temp = seq[k]
        seq[k] = seq[index]
        seq[index] = temp

    # counting the number of hits
    cnt = 0
    for i in range(1, len(seq)):
        if(seq[i] == i):
            cnt += 1
    v.append(cnt)

expectation = np.mean(v)
variance = np.var(v) 
print(expectation, variance)


1.00454 1.0075193884000002


##### Empirical estimates
- Expectation: 1
- Variance: 1

### Q5

In [165]:
def get_random_permutation_b(n):
    output = [[1]]
    print(output[0])
    for i in range(2, n+1):
        # previous sequence
        seq = output[-1]
        # perm is of length i
        perm = seq + [i]
        u = np.random.rand()
        index = int(i*u) + 1

        # swapping the 2 numbers
        temp = perm[index-1]
        perm[index-1] = perm[i-1]
        perm[i-1] = temp

        print(perm)
        output.append(perm)
    return output

In [174]:
output = get_random_permutation_b(6)

[1]
[1, 2]
[3, 2, 1]
[3, 4, 1, 2]
[3, 5, 1, 2, 4]
[3, 5, 1, 2, 4, 6]


### Q6

##### Method 1

In [198]:
# u = np.random.rand()
u = random_number_sequence[0]
prob = 0.8
len_seq = 25
output = np.zeros(len_seq)
for i in range(len_seq):
    if(u<=prob):
        val = 1
    else:
        val = 0
    output[i] = val
    # updating u for the next sampling
    if(u<=prob):
        u = u/prob
    else:
        u = (u-prob)/(1-prob)
output

array([1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1., 0., 1., 0., 1., 1., 0.,
       1., 1., 1., 1., 1., 1., 1., 1.])

In [199]:
len(output[output==1])

21

##### just 1 random number was needed!

##### Method 2

In [230]:
prob = 0.80
i = 0
output = []
while(True):
    u = random_number_sequence[i]
    x = int(math.log(u)/math.log(prob)) + 1
    for _ in range(x-1):
        output.append(1)
        if(len(output) == 25):
            break
    if(len(output) == 25):
        break
    output.append(0)
    if(len(output) == 25):
        break
    i += 1
output

[1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1]

##### Since there are 6 sequences of 1 (or starting with 1), 6 random numbers were needed! 

### Q7

In [164]:
num_trials = int(2e5)
ans = 0
for _ in range(num_trials):
    # cnt represents the number of times we throw the pair of dice; 
    #   number of throws required to get all possible sums atleast once
    cnt = 0
    # s stores only the distinct values; so once s has all the elements from 2 to 12 
    #   (11 possible values), we are guaranteed to have each sum atleast once
    #   Note that no other value (of the sum) is possible; 
    #   so len(s) = 11 <=> all possible sums (from 2 to 12) occured atleast once 
    s = set([])
    while(len(s) != 11):
    # while(len(s) != 6):
        # simulate throwing of 2 dice
        cnt += 1
        # get the value
        u1 = np.random.rand()
        v1 = int(6*u1) + 1
        u2 = np.random.rand()
        v2 = int(6*u2) + 1
        # v2 = 0
        s.add(v1+v2)
    ans += cnt
ans = ans/num_trials
print(ans)

61.18592


##### The estimate (expected number of dice rolls) seems to be 61.2 => upper/ceilinged to 62